### PR TITLE
SNOW-833713 snow sql performance issue with ocsp validation in 1 2 26

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-833713

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [x] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

 - lazy evaluation of the module level ocsp cache variable, so that if the connector is in in_secure mode, the ocsp cache doesn't get initialized at all
 - reduce the times to dump cache to disks, if there's no change in the cache, we do not need to re-write file

### CODE REVIEW

**Comment by dataders** [comment]
was this an issue surfaced as a result of #1582?

**Comment by sfc-gh-aling** [comment]
@dataders no, they two are irrelevant.
#1582 is to clean up temp file on windows.
this PR tries to improve the OCSP cache performance.

File: src/snowflake/connector/cache.py [hunk 1]
```diff
@@ -58,6 +58,9 @@ def __init__(
         self._cache: dict[K, CacheEntry[V]] = {}
         self._lock = Lock()
         self._reset_telemetry()
+        # indicate whether the cache is modified or not, this variable is for
+        # SFDictFileCache to determine whether to dump cache to file when _save is called
+        self._cache_modified = False
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:63]: We should move all of these to `SFDictFileCache`, this class has no use for this information!

File: src/snowflake/connector/cache.py [hunk 2]
```diff
@@ -494,7 +514,8 @@ def __getitem__(self, k: K) -> V:
 
     def _setitem(self, k: K, v: V) -> None:
         super()._setitem(k, v)
-        self._save_if_should()
+        if self._try_saving_when_set_item:
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:517]: I think this check could be moved into `_save_if_should`

File: src/snowflake/connector/cache.py [hunk 3]
```diff
@@ -511,9 +532,12 @@ def _load(self) -> bool:
             logger.debug("Fail to read cache from disk due to error: %s", e)
             return False
 
-    def _save(self, load_first: bool = True) -> bool:
+    def _save(self, load_first: bool = True, force_flush: bool = False) -> bool:
         """Save cache to disk if possible, returns whether it was able to save."""
         self._clear_expired_entries()
+        if not self._cache_modified and not force_flush:
+            # cache is not updated, so there is no need to dump cache to file, we just return
+            return True
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:540]: It didn't save, that should be `False`
```suggestion
            return False
```

File: src/snowflake/connector/cache.py [hunk 4]
```diff
@@ -572,6 +598,10 @@ def _save_if_should(self) -> bool:
             return self._save()
         return False
 
+    def save(self, load_first: bool = True):
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/cache.py:601]: I don't think that we should expose this

File: src/snowflake/connector/ocsp_snowflake.py [hunk 5]
```diff
@@ -74,38 +74,50 @@ class OCSPResponseValidationResult(NamedTuple):
...
-    OCSP_RESPONSE_VALIDATION_CACHE: SFDictCache[
-        tuple[bytes, bytes, bytes],
-        OCSPResponseValidationResult,
-    ] = SFDictCache(
-        entry_lifetime=constants.DAY_IN_SECONDS,
-    )
+# use _get_ocsp_response_validation_cache instead of the module variable for lazy evaluation
+OCSP_RESPONSE_VALIDATION_CACHE: SFDictFileCache[
+    tuple[bytes, bytes, bytes], OCSPResponseValidationResult
+] | SFDictCache[tuple[bytes, bytes, bytes], OCSPResponseValidationResult] | None = None
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:80]: Drop the `None` here and make it be the return value of `_get_ocsp_response_validation_cache`
```suggestion
] | SFDictCache[tuple[bytes, bytes, bytes], OCSPResponseValidationResult]  = _get_ocsp_response_validation_cache()
```
This way you can drop the types and push them onto the function's return type-hint as well.

File: src/snowflake/connector/ocsp_snowflake.py [hunk 6]
```diff
@@ -74,38 +74,50 @@ class OCSPResponseValidationResult(NamedTuple):
...
+
+
+def _get_ocsp_response_validation_cache():
+    global OCSP_RESPONSE_VALIDATION_CACHE
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:84]: This is an invisible side-effect of this function and usually a bad idea!

File: src/snowflake/connector/ocsp_snowflake.py [hunk 7]
```diff
@@ -585,9 +599,10 @@ def update_file(ocsp) -> None:
             1. file for OCSP_RESPONSE_VALIDATION_CACHE which keeps ocsp response validation result
             2. ocsp_response_cache.json, the file in the same format as the one downloaded from snowflake cache service
         """
+        ocsp_response_validation_cache = _get_ocsp_response_validation_cache()
         if OCSPCache.CACHE_UPDATED:
-            if isinstance(OCSP_RESPONSE_VALIDATION_CACHE, SFDictFileCache):
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:589]: The rest of this file should not change. You should be operating on the global cache every time. We resolve it once and then keep using that one!

File: test/extras/run.py [hunk 8]
```diff
@@ -27,7 +27,7 @@
         )
         sub_process.check_returncode()
         ocsp_cache_dir_path = pathlib.Path(
-            snowflake.connector.ocsp_snowflake.OCSP_RESPONSE_VALIDATION_CACHE.file_path
+            snowflake.connector.ocsp_snowflake._get_ocsp_response_validation_cache().file_path
```
Comment (sfc-gh-mkeller) [inline test/extras/run.py:30]: Shouldn't change

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: src/snowflake/connector/cache.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:517]: `_save_if_should` is also used by `clear_expired_entries`, so I do want `_try_saving_when_set_item` to impact the behavior of `clear_expired_entries`. or we could introduce another parameter in `_save_if_should` to control the behavior

File: src/snowflake/connector/cache.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:63]: yep, you're right!

File: src/snowflake/connector/ocsp_snowflake.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py:80]: I want this OCSP_RESPONSE_VALIDATION_CACHE to be lazy instantiated. calling the function conflicts my purpose...

File: src/snowflake/connector/ocsp_snowflake.py (hunk 6 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py:84]: oh, could you elaborate on this? I might be unfamiliar with the potential pitfalls here

File: src/snowflake/connector/cache.py (hunk 4 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:601]: I will remove this

File: src/snowflake/connector/cache.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py]: updated

File: src/snowflake/connector/ocsp_snowflake.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: I have revert the change to lazy initialization, by simply moving the import into the if condition in `ssl_wrap_socket`:
```python
    if FEATURE_OCSP_MODE != OCSPMode.INSECURE:
        from .ocsp_asn1crypto import SnowflakeOCSPAsn1Crypto as SFOCSP
```

File: src/snowflake/connector/ocsp_snowflake.py (hunk 7 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: see comment above, change reverted

File: test/extras/run.py (hunk 8 above)
Comment (sfc-gh-aling) [reply test/extras/run.py]: see comment above, change reverted

File: src/snowflake/connector/ocsp_snowflake.py (hunk 6 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: see comment above, change reverted

File: src/snowflake/connector/ocsp_snowflake.py [hunk 9]
```diff
@@ -364,7 +365,8 @@ def download_cache_from_server(self, ocsp):
                 # len(OCSP_RESPONSE_VALIDATION_CACHE) is thread-safe, however, we do not want to
                 # block for logging purpose, thus using len(OCSP_RESPONSE_VALIDATION_CACHE._cache) here.
                 logger.debug(
-                    "# of certificates: %s", len(OCSP_RESPONSE_VALIDATION_CACHE._cache)
+                    "# of certificates: %s",
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:368]: `%s` is for strings, `%i`, or `%u` might be more appropiate
```suggestion
                    "# of certificates: %u",
```

File: src/snowflake/connector/ocsp_snowflake.py [hunk 10]
```diff
@@ -524,7 +526,8 @@ def reset_ocsp_response_cache_uri(ocsp_response_cache_uri) -> None:
         # len(OCSP_RESPONSE_VALIDATION_CACHE) is thread-safe, however, we do not want to
         # block for logging purpose, thus using len(OCSP_RESPONSE_VALIDATION_CACHE._cache) here.
         logger.debug(
-            "OCSP_VALIDATION_CACHE size: %s", len(OCSP_RESPONSE_VALIDATION_CACHE._cache)
+            "OCSP_VALIDATION_CACHE size: %s",
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/ocsp_snowflake.py:529]: `%s` is for strings, `%i`, or `%u` might be more appropiate
```suggestion
            "OCSP_VALIDATION_CACHE size: %u",
```

File: test/integ/test_connection.py [hunk 11]
```diff
@@ -1276,3 +1276,21 @@ def test_server_session_keep_alive(conn_cnx):
     with conn_cnx() as conn:
         conn.rest.delete_session = mock_delete_session
     mock_delete_session.assert_called_once()
+
+
+@pytest.mark.skipolddriver
+def test_ocsp_mode_in_secure(db_parameters, is_public_test, caplog):
```
Comment (sfc-gh-mkeller) [inline test/integ/test_connection.py:1282]: Use `conn_cnx` instead of `db_parameters`

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: src/snowflake/connector/ocsp_snowflake.py (hunk 9 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/ocsp_snowflake.py]: updated

**Pull request merged**